### PR TITLE
improve: feedback and logging for dst remove and id remove

### PR DESCRIPTION
**Pull request by kimskimchi**
## Summary

**Major**
- introduces force flag to `destinations remove` and `identities remove`
   ```
   $ infra dst rm abc
   No destinations named [abc].
   $ infra echo $?
   1

   $ infra dst rm abc --force
   $ echo $?
   0
   ```
- `destinations remove` and `identities remove` starts using CLI Error 

**Minor**
- updates logging of CLI Error to not be two lines, but rather one long one 
  ```
  old: 
    Internal error: 
    some error..
  now: 
    Internal error: some error...
  ```
- adds logging for `destinations [list|remove]`, and `identities remove`
- updates docs for `--force` flag for `destinations remove` and `identities remove`
- fixes tests for delete (api get and delete are now working properly)

## Checklist

- [x] Wrote appropriate unit tests --> **[updated tests]**
- [ ] Considered security implications of the change
- [x] Updated associated docs where necessary  --> **[make docs]**
- [x] Updated associated configuration where necessary
- [x] Is change backwards compatible? --> **[yes, will return 0 on not found if `--force` is specified]**
- [x] Nothing sensitive logged --> **[logs input, no sensitive info]**
- [ ] Considered data migrations for smooth upgrades

## Related Issues

<!--
Link any related issues. Each issue should be on
its own line. For example:

Resolves #1234
Resolves #4321
--
Related #1858

### CODE REVIEW

File: internal/cmd/destinations.go [hunk 1]
```diff
@@ -40,6 +41,7 @@ func newDestinationsListCmd(cli *CLI) *cobra.Command {
 				return err
 			}
 
+			logging.S.Debug("call server: list destinations")
```
Comment (BruceMacD) [inline internal/cmd/destinations.go:44]: Any info about the current runtime we can add to these debug logs that might help give even more context? Not needed, just something I thought of.

File: main.go [hunk 2]
```diff
@@ -3,19 +3,28 @@ package main
...
 
 func main() {
+
 	if err := cmd.Run(context.Background(), os.Args[1:]...); err != nil {
-		if !errors.Is(err, terminal.InterruptErr) {
-			logging.S.Error(err.Error())
+		var userErr cmd.Error
+		switch {
+		case errors.Is(err, terminal.InterruptErr):
+			// print nothing, the user initiated the exit
```
Comment (BruceMacD) [inline main.go:20]: should we debug log this too?

File: internal/cmd/identities.go [hunk 3]
```diff
@@ -175,6 +189,8 @@ $ infra identities remove janedoe@example.com`,
 		},
 	}
 
+	cmd.Flags().BoolVar(&force, "force", false, "Exit successfully destination not found")
```
Comment (BruceMacD) [inline internal/cmd/identities.go:192]:
```suggestion
	cmd.Flags().BoolVar(&force, "force", false, "Exit successfully when destination not found")
```

File: internal/cmd/destinations.go (hunk 1 above)
Comment (kimskimchi) [reply internal/cmd/destinations.go:44]: I have been struggling a bit attempting to error handle / log  'apiclient'/contexts, etc, since there is #1400 that needs to be addressed first

File: internal/cmd/destinations.go [hunk 4]
```diff
@@ -82,16 +87,24 @@ func newDestinationsRemoveCmd() *cobra.Command {
...
 			if err != nil {
 				return err
 			}
 
 			if destinations.Count == 0 {
-				return fmt.Errorf("no destinations named %s.", args[0])
+				if force {
+					return nil
+				}
+				return Error{
```
Comment (pdevine) [inline internal/cmd/destinations.go:100]: We haven't switched all of the other errors over to using this style of Error{} object. Do you want to create this one in the old fmt.Errorf() format and then convert everything over at once? That way we won't have skew between the other styles of errors.

File: main.go [hunk 5]
```diff
@@ -12,10 +13,19 @@ import (
 )
 
 func main() {
+
 	if err := cmd.Run(context.Background(), os.Args[1:]...); err != nil {
-		if !errors.Is(err, terminal.InterruptErr) {
-			logging.S.Error(err.Error())
+		var userErr cmd.Error
```
Comment (pdevine) [inline main.go:18]: I think this can go in as a separate PR

**Review by pdevine — CHANGES_REQUESTED** [review]
Overall looks fine. It'd be nice to split out the errors part from this PR and put them in separately though just to keep local consistency.

File: main.go (hunk 5 above)
Comment (kimskimchi) [reply main.go]: done~

File: internal/access/identity_test.go [hunk 6]
```diff
@@ -1,100 +0,0 @@
-package access
```
Comment (ssoroka) [inline internal/access/identity_test.go:1]: not sure why this is here.

File: internal/cmd/destinations.go [hunk 7]
```diff
@@ -58,10 +60,11 @@ func newDestinationsListCmd(cli *CLI) *cobra.Command {
 				})
 			}
 
+			logging.S.Debug("print destinations")
```
Comment (ssoroka) [inline internal/cmd/destinations.go:63]: doesn't seem useful

File: internal/cmd/destinations.go (hunk 4 above)
Comment (ssoroka) [inline internal/cmd/destinations.go:100]: is this a user-facing error? I think perhaps we should change the name of the type to something more descriptive? at the moment it seems like `Error` is appropriate to use for all errors, but I'm assuming it's only appropriate for errors where you want to customize the text to be something the user sees and acts on. This probably differs from other errors in that these shouldn't be wrapped, prefixed, or displayed with stack traces, etc. thoughts?

File: internal/cmd/error.go [hunk 8]
```diff
@@ -18,14 +18,14 @@ type Error struct {
 func (e Error) Error() string {
 	if e.OriginalError != nil {
 		if len(e.Message) == 0 {
-			return fmt.Sprintf("Internal error:\n%v", e.OriginalError)
+			return fmt.Sprintf("Internal error: %v", e.OriginalError)
```
Comment (ssoroka) [inline internal/cmd/error.go:21]: `Internal error` doesn't seem accurate here. The Error type is for user-facing errors is it not?

**Review by ssoroka — CHANGES_REQUESTED** [review]
The `logging.S.Debug` statements don't seem that useful for users debugging what's going wrong with a command; instead they seem to be of the breakpoint style that a developer would use while writing the program. Users instead want to see hidden server responses, warnings, etc.

File: internal/cmd/destinations.go (hunk 7 above)
Comment (kimskimchi) [reply internal/cmd/destinations.go:63]: removed

File: internal/cmd/destinations.go (hunk 4 above)
Comment (kimskimchi) [reply internal/cmd/destinations.go:100]: We initially named it a "userfacing error" or some sort, but then after some discussion, a few of us were okay with naming it "error", since if we named it "UserFacingError" we don't need "Error" right now anyways (in the future we could rename if it becomes more robust and we need to split them)

File: internal/cmd/error.go (hunk 8 above)
Comment (kimskimchi) [reply internal/cmd/error.go:21]: The purpose of the Error is to allow for user-facing output, so any errors (internal or developer specified) can all be formatted consistently. Basically to wrap any errors (or create any) that can be printed on the CLI nicely

File: internal/cmd/users.go [hunk 9]
```diff
@@ -156,12 +158,24 @@ $ infra users remove janedoe@example.com`,
...
+			if users.Count == 0 {
+				if force {
+					return nil
+				}
+				return Error{
+					Message: fmt.Sprintf("No users named [%s].", name),
+				}
+			}
+
+			logging.S.Debug("deleting %d users named [%s]...", users.Count, name)
```
[the remark this answers is not in the document]
Comment (kimskimchi) [reply internal/cmd/users.go:176]: I actually agree, but this was the behavior that was requested (by I believe Jeff). I initially proposed that we error out and provide a message, to panic or output "multiple users found", but the behavior that was requested was to just ignore and assume API is always doing the correct thing.

File: internal/cmd/users_test.go [hunk 10]
```diff
@@ -142,10 +156,27 @@ func TestUsersCmd(t *testing.T) {
 	t.Run("removes only the specified user", func(t *testing.T) {
 		modifiedUsers := setup(t)
 		ctx := context.Background()
-		err := Run(ctx, "users", "remove", "to-delete-user@example.com")
+		err := Run(ctx, "users", "add", "to-delete-user@example.com")
```
[the remark this answers is not in the document]
Comment (kimskimchi) [reply internal/cmd/users_test.go:168]: done

File: internal/access/identity_test.go (hunk 6 above)
Comment (kimskimchi) [reply internal/access/identity_test.go]: done

**Comment by mxyng** [comment]
The force remove flag has been merged in #2042
